modules/codegen/src/main/scala/com/twilio/guardrail/generators/circe/protocol/CirceProtocolGenerator.scala

```diff
@@ -284,76 +308,76 @@ object CirceProtocolGenerator {
-    def apply[T](term: ProtocolSupportTerm[T]): Target[T] = term match {
-      case ExtractConcreteTypes(definitions) =>
-        Target.getGeneratorSettings.flatMap { implicit gs =>
-          for {
-            entries <- definitions.traverse {
-              case (clsName, impl: ModelImpl) if (Option(impl.getProperties()).isDefined || Option(impl.getEnum()).isDefined) =>
-                Target.pure((clsName, SwaggerUtil.Resolved(Type.Name(clsName), None, None): SwaggerUtil.ResolvedType))
-              case (clsName, definition) =>
-                SwaggerUtil
-                  .modelMetaType(definition)
-                  .map(x => (clsName, x))
-            }
-            result <- SwaggerUtil.ResolvedType.resolve_(entries)
-          } yield
-            result.map {
-              case (clsName, SwaggerUtil.Resolved(tpe, _, _)) =>
-                PropMeta(clsName, tpe)
-            }
-        }
+  object PolyProtocolTermInterp extends (PolyProtocolTerm ~> Target) {
+    override def apply[A](fa: PolyProtocolTerm[A]): Target[A] = fa match {
+      case RenderADTCompanion(clsName, discriminator, encoder, decoder) =>
+        val code = q"""object ${Term.Name(clsName)} {
+             val discriminator:String = ${Lit.String(discriminator)}
+            ..${List(encoder, decoder)}
+          }
+          """
+        Target.pure(code)
 
-      case ProtocolImports() =>
-        Target.pure(
-          List(
-            q"import io.circe._",
-            q"import io.circe.syntax._",
-            q"import io.circe.generic.semiauto._"
-          )
+      case DecodeADT(clsName, children) =>
+        val childrenCases = children.map(
+          child => p"case ${Lit.String(child)} => c.as[${Type.Name(child)}]"
         )
+        val code =
+          q"""implicit val decoder: Decoder[${Type.Name(clsName)}] = Decoder.instance(c =>
+                 c.downField(discriminator).as[String].flatMap {
+                   ..case $childrenCases
+                 }
+            )"""
+        Target.pure(code)
+
+      case EncodeADT(clsName, children) =>
+        val childrenCases = children.map(
+          child => p"case e:${Type.Name(child)} => e.asJsonObject.add(discriminator, ${Lit.String(child)}.asJson).asJson"
+        )
+        val code =
+          q"""implicit val encoder: Encoder[${Type.Name(clsName)}] = Encoder.instance {
+              ..case $childrenCases
+          }"""
+        Target.pure(code)
+
+      case RenderSealedTrait(className, terms, discriminator, parents) =>
+        val parentNameOpt = parents.headOption.map(_.clsName)
+        val testTerms = terms
+          .filter(_.name.value != discriminator)
+          .map { t =>
+            val tpe: Type = t.decltpe
+              .flatMap({
+                case tpe: Type => Some(tpe)
+                case x =>
+                  println(s"Unsure how to map ${x.structure}, please report this bug!")
+                  None
+              })
+              .get
```

Comment:
Please `traverse` in order to use `Target.error` instead of throwing an exception

Reply:
(Previous discussion: https://github.com/twilio/guardrail/pull/98#discussion_r221834486)
Despite `None` being impossible, there's still a `println` there. As `traverse` would make using `Target.error` very easy, I don't see a good reason to use `.get` even if it is impossible to fail.